taichi/rhi/metal/metal_device.mm

```diff
@@ -943,111 +1279,75 @@ DeviceCapabilityConfig collect_metal_device_caps(MTLDevice_id mtl_device) {
       return nullptr;
     }
 
-    const std::string new_entry_point_name =
-        src[i].stage == PipelineStageType::fragment
-            ? std::string(kMetalFragFunctionName)
-            : std::string(kMetalVertFunctionName);
-
-    // Fragment and vertex function names must be different.
-    // If spirv-cross has a method to let you set the emitted entry point's
-    // name, that would be nice, but I could not find anything in the docs.
-    // So for now just using std's regex_replace().
-    std::regex entry_point_regex("main0");
-    msl_source +=
-        std::regex_replace(msl, entry_point_regex, new_entry_point_name);
-  }
-
-  // Compile MSL source
-  MTLLibrary_id mtl_library = get_mtl_library(msl_source);
+    // Find mapping of GLSL binding to MSL index
+    spirv_cross::ShaderResources shader_res = compiler.get_shader_resources(
+        compiler.get_active_interface_variables());
+
+    // TODO: Refactor this ugly, hard-to-read double for-loop.
+    for (int i = 0; i < 4; i++) {
+      spirv_cross::SmallVector<spirv_cross::Resource> *resource_list =
+          &(shader_res.uniform_buffers);
+      if (i == 1)
+        resource_list = &(shader_res.storage_buffers);
+      if (i == 2)
+        resource_list = &(shader_res.sampled_images);
+      if (i == 3)
+        resource_list = &(shader_res.storage_images);
```

Comment:
one way to refactor this could be making the inner loop an lambda expression that takes a pointer to a particular resource_list with a boolean flag to control whether there is a sampler index as well